test/bitonic_sort_test.dart

```diff
@@ -0,0 +1,479 @@
+            BitonicSort(clk, reset, toSort: toSort, name: 'top_level');
+        await topMod.build();
+
+        reset.inject(0);
+
+        toSortRes.sort();
+        Simulator.registerAction(100, () {
+          for (var i = 0; i < topMod.sorted.length; i++) {
+            expect(topMod.sorted[i].value.toInt(), toSortRes[i]);
+          }
+        });
+
+        Simulator.setMaxSimTime(100);
+
+        await Simulator.run();
+      });
+
+      test(
+          'should return the sorted results in ascending order given '
+          'the inputs consists of duplicates.', () async {
+        const dataWidth = 8;
+
+        final clk = SimpleClockGenerator(10).clk;
+        final reset = Logic(name: 'reset');
+
+        final toSort = <Logic>[
+          Const(1, width: dataWidth),
+          Const(7, width: dataWidth),
+          Const(1, width: dataWidth),
+          Const(8, width: dataWidth),
+        ];
+
+        final topMod =
+            BitonicSort(clk, reset, toSort: toSort, name: 'top_level');
+        await topMod.build();
+
+        reset.inject(0);
+
+        final toSortRes = [1, 1, 7, 8];
+
+        Simulator.registerAction(100, () {
+          for (var i = 0; i < topMod.sorted.length; i++) {
+            expect(topMod.sorted[i].value.toInt(), toSortRes[i]);
+          }
+        });
+
+        Simulator.setMaxSimTime(100);
+
+        await Simulator.run();
+      });
+
+      test(
+          'should return sorted results after latency of the'
+          ' sorting completed for all random numbers in ascending, pipeline.',
+          () async {
+        const dataWidth = 10;
+
+        final clk = SimpleClockGenerator(10).clk;
+        final reset = Logic(name: 'reset');
+
+        final a = Logic(name: 'in_1', width: dataWidth);
+        final b = Logic(name: 'in_2', width: dataWidth);
+        final c = Logic(name: 'in_3', width: dataWidth);
+        final d = Logic(name: 'in_4', width: dataWidth);
+        final toSort = <Logic>[a, b, c, d];
+
+        final topMod =
+            BitonicSort(clk, reset, toSort: toSort, name: 'top_level');
+
+        await topMod.build();
+
+        Simulator.setMaxSimTime(1000);
+        unawaited(Simulator.run());
+
+        Future<void> waitCycles(int numCycles) async {
+          for (var i = 0; i < numCycles; i++) {
+            await clk.nextPosedge;
+          }
+        }
+
+        final inputs = List.generate(
+            10, (index) => List.generate(4, (index) => Random().nextInt(10)));
+
+        for (final input in inputs) {
+          a.put(input[0]);
+          b.put(input[1]);
+          c.put(input[2]);
+          d.put(input[3]);
+
+          input.sort();
+
+          await waitCycles(topMod.latency).then((value) {
+            for (var i = 0; i < topMod.sorted.length; i++) {
+              expect(topMod.sorted[i].value.toInt(), input[i]);
+            }
+          });
+
+          await clk.nextNegedge;
+        }
+
+        await Simulator.simulationEnded;
```

Comment:
similar feedback as in the multiplier, end the simulation explicitly, don't `await` the `waitCycles`, ensure the `expect` ran the right number of times